packages/data-context/src/data/ProjectLifecycleManager.ts

```diff
@@ -240,12 +241,21 @@ export class ProjectLifecycleManager {
           finalConfig.baseUrl = `http://localhost:${devServerOptions?.port}`
         }
 
+        const restartOnChange = validateNeedToRestartOnChange(this._cachedFullConfig, finalConfig)
+
         this._cachedFullConfig = finalConfig
 
         // This happens automatically with openProjectCreate in run mode
         if (!this.ctx.isRunMode) {
           if (!this._initializedProject) {
             this._initializedProject = await this.ctx.actions.project.initializeActiveProject({})
+          } else if (restartOnChange.server) {
+            this.ctx.project.setRelaunchBrowser(true)
+            this._initializedProject = await this.ctx.actions.project.initializeActiveProject({})
+          } else if (restartOnChange.browser) {
+            this.ctx.project.setRelaunchBrowser(true)
+            await this.ctx.actions.browser.closeBrowser()
+            await this.ctx.actions.browser.focusActiveBrowserWindow()
```

Comment:
What happens if they haven't opened the browser yet, but they make a config change, does this end up launching the browser?

Reply:
Good catch! I'm gonna update it to handle it

Reply:
[2303fa5](https://github.com/cypress-io/cypress/pull/21327/commits/2303fa559703d5cbb54e8a2aacb0382f79f87abb)